File: src/export.ml
type t

external add_function_export : Module.t -> string -> string -> t
  = "caml_binaryen_add_function_export"
(** Module, internal name, external name. *)

external add_table_export : Module.t -> string -> string -> t
  = "caml_binaryen_add_table_export"
(** Module, internal name, external name. *)

external add_memory_export : Module.t -> string -> string -> t
  = "caml_binaryen_add_memory_export"
(** Module, internal name, external name. *)

external add_global_export : Module.t -> string -> string -> t
  = "caml_binaryen_add_global_export"
(** Module, internal name, external name. *)

external add_tag_export : Module.t -> string -> string -> t
  = "caml_binaryen_add_tag_export"
(** Module, internal name, external name. *)

external get_export : Module.t -> string -> t = "caml_binaryen_get_export"

external remove_export : Module.t -> string -> unit
  = "caml_binaryen_remove_export"
(** Module, external name. *)

external get_num_exports : Module.t -> int = "caml_binaryen_get_num_exports"

external get_export_by_index : Module.t -> int -> t
  = "caml_binaryen_get_export_by_index"

external get_name : t -> string = "caml_binaryen_export_get_name"
external get_value : t -> string = "caml_binaryen_export_get_value"
external external_function : unit -> int = "caml_binaryen_external_function"

let external_function = external_function ()

external external_table : unit -> int = "caml_binaryen_external_table"

let external_table = external_table ()

external external_memory : unit -> int = "caml_binaryen_external_memory"

let external_memory = external_memory ()

external external_global : unit -> int = "caml_binaryen_external_global"

let external_global = external_global ()

external external_tag : unit -> int = "caml_binaryen_external_tag"

let external_tag = external_tag ()

external export_get_kind : t -> int = "caml_binaryen_export_get_kind"

File: src/export.js
//Provides: caml_binaryen_add_function_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_function_export(
  wasm_mod,
  internal_name,
  external_name
) {
  return wasm_mod.addFunctionExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_table_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_table_export(
  wasm_mod,
  internal_name,
  external_name
) {
  return wasm_mod.addTableExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_memory_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_memory_export(
  wasm_mod,
  internal_name,
  external_name
) {
  return wasm_mod.addMemoryExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_add_global_export
//Requires: caml_jsstring_of_string
function caml_binaryen_add_global_export(
  wasm_mod,
  internal_name,
  external_name
) {
  return wasm_mod.addGlobalExport(
    caml_jsstring_of_string(internal_name),
    caml_jsstring_of_string(external_name)
  );
}

//Provides: caml_binaryen_get_export
//Requires: caml_jsstring_of_string
function caml_binaryen_get_export(wasm_mod, external_name) {
  return wasm_mod.getExport(caml_jsstring_of_string(external_name));
}

//Provides: caml_binaryen_remove_export
//Requires: caml_jsstring_of_string
function caml_binaryen_remove_export(wasm_mod, external_name) {
  return wasm_mod.removeExport(caml_jsstring_of_string(external_name));
}

//Provides: caml_binaryen_get_num_exports
function caml_binaryen_get_num_exports(wasm_mod) {
  return wasm_mod.getNumExports();
}

//Provides: caml_binaryen_get_export_by_index
function caml_binaryen_get_export_by_index(wasm_mod, index) {
  return wasm_mod.getExportByIndex(index);
}

//Provides: caml_binaryen_export_get_name
//Requires: Binaryen
//Requires: caml_string_of_jsstring
function caml_binaryen_export_get_name(exp) {
  var export_info = Binaryen.getExportInfo(exp);
  return caml_string_of_jsstring(export_info.name);
}

//Provides: caml_binaryen_export_get_value
//Requires: Binaryen
//Requires: caml_string_of_jsstring
function caml_binaryen_export_get_value(exp) {
  var export_info = Binaryen.getExportInfo(exp);
  return caml_string_of_jsstring(export_info.value);
}

//Provides: caml_binaryen_external_function
//Requires: Binaryen
function caml_binaryen_external_function() {
  return Binaryen._BinaryenExternalFunction();
}

//Provides: caml_binaryen_external_table
//Requires: Binaryen
function caml_binaryen_external_table() {
  return Binaryen._BinaryenExternalTable();
}

//Provides: caml_binaryen_external_memory
//Requires: Binaryen
function caml_binaryen_external_memory() {
  return Binaryen._BinaryenExternalMemory();
}

//Provides: caml_binaryen_external_global
//Requires: Binaryen
function caml_binaryen_external_global() {
  return Binaryen._BinaryenExternalGlobal();
}

//Provides: caml_binaryen_external_tag
//Requires: Binaryen
function caml_binaryen_external_tag() {
  return Binaryen._BinaryenExternalTag();
}

//Provides: caml_binaryen_export_get_kind
//Requires: Binaryen
function caml_binaryen_export_get_kind(exp) {
  var export_info = Binaryen.getExportInfo(exp);
  return export_info.kind;
}
